src/main/java/dev/dbos/transact/database/WorkflowDAO.java

```diff
@@ -110,38 +110,31 @@ public WorkflowInitResult initWorkflowStatus(
         final int attempts = resRow.getRecoveryAttempts();
         if (maxRetries != null && attempts > maxRetries + 1) {
 
-          UpdateWorkflowOptions options = new UpdateWorkflowOptions();
-          options.setWhereStatus(WorkflowState.PENDING.toString());
-          options.setThrowOnFailure(false);
-
-          updateWorkflowStatus(
-              connection,
-              initStatus.getWorkflowUUID(),
-              WorkflowState.RETRIES_EXCEEDED.toString(),
-              options);
-          throw new DBOSDeadLetterQueueException(initStatus.getWorkflowUUID(), maxRetries);
+          var sql =
+              """
+              UPDATE %s.workflow_status
+              SET status = ?, deduplication_id = NULL, started_at_epoch_ms = NULL, queue_name = NULL
+              WHERE workflow_uuid = ? AND status = ?
+              """
+                  .formatted(Constants.DB_SCHEMA);
+
+          try (PreparedStatement stmt = connection.prepareStatement(sql)) {
+            stmt.setString(1, WorkflowState.MAX_RECOVERY_ATTEMPTS_EXCEEDED.toString());
+            stmt.setString(2, initStatus.getWorkflowUUID());
+            stmt.setString(3, WorkflowState.PENDING.toString());
+
+            stmt.executeUpdate();
+          }
+
+          throw new DBOSMaxRecoveryAttemptsExceededException(
+              initStatus.getWorkflowUUID(), maxRetries);
         }
 
-        connection.commit(); // Commit transaction on success
         return new WorkflowInitResult(
             initStatus.getWorkflowUUID(), resRow.getStatus(), resRow.getWorkflowDeadlineEpochMs());
 
-      } catch (SQLException e) {
-
-        try {
-          connection.rollback();
-        } catch (SQLException rollbackEx) {
-          logger.error("Rollback failed", rollbackEx);
-        }
-        throw e; // Re-throw the original SQLException
-      } catch (DBOSConflictingWorkflowException | DBOSDeadLetterQueueException e) {
-        // Rollback for custom business exceptions
-        try {
-          connection.rollback();
-        } catch (SQLException rollbackEx) {
-          logger.error("Rollback failed", rollbackEx);
-        }
-        throw e; // Re-throw the custom exception
+      } finally {
+        connection.commit();
```

Comment:
What happens if a database exception is thrown during the transaction, for example because of a deduplication key violation?

Reply:
I'll add a test to be sure, but `insertWorkflowStatus` throws a DBOSQueueDuplicatedException if it detects a unique contraint violation (aka `23505`). THe tx will commit, but it won't have made a change. This is the same logic as the TS implementation (Python doesn't use a tx in init_workflow)